grudge/discretization.py

```diff
@@ -636,17 +563,166 @@ def normal(self, dd):
+    """
+    if not isinstance(discr_context, Mesh):
+        raise NotImplementedError(
+            "Currently `discr_context` must be a meshmode Mesh object"
+        )
+    mesh = discr_context
 
-class DGDiscretizationWithBoundaries(DiscretizationCollection):
-    def __init__(self, *args, **kwargs):
-        warn("DGDiscretizationWithBoundaries is deprecated and will go away "
-                "in 2022. Use DiscretizationCollection instead.",
-                DeprecationWarning, stacklevel=2)
+    from meshmode.discretization.poly_element import \
+        default_simplex_group_factory
 
-        super().__init__(*args, **kwargs)
+    if discr_tag_to_group_factory is None:
+        if order is None:
+            raise TypeError(
+                "one of 'order' and 'discr_tag_to_group_factory' must be given"
+            )
+
+        discr_tag_to_group_factory = {
+            DISCR_TAG_BASE: default_simplex_group_factory(base_dim=mesh.dim,
+                                                          order=order)
+        }
+    else:
+        if order is not None:
+            discr_tag_to_group_factory = discr_tag_to_group_factory.copy()
+            if DISCR_TAG_BASE in discr_tag_to_group_factory:
+                raise ValueError(
+                    "if 'order' is given, 'discr_tag_to_group_factory' must "
+                    "not have a key of DISCR_TAG_BASE"
+                )
+
+            discr_tag_to_group_factory[DISCR_TAG_BASE] = \
+                default_simplex_group_factory(base_dim=mesh.dim, order=order)
+
+    # Modal discr should always comes from the base discretization
+    discr_tag_to_group_factory[DISCR_TAG_MODAL] = \
+        _generate_modal_group_factory(
+            discr_tag_to_group_factory[DISCR_TAG_BASE]
+        )
+
+    # Define the base and modal discretization
+    from meshmode.discretization import Discretization
+
+    volume_discr = Discretization(
+        array_context, mesh,
+        discr_tag_to_group_factory[DISCR_TAG_BASE]
+    )
+
+    modal_vol_discr = Discretization(
+        array_context, mesh,
+        discr_tag_to_group_factory[DISCR_TAG_MODAL]
+    )
+
+    discr_from_dd = {DD_VOLUME: volume_discr,
+                     DD_VOLUME_MODAL: modal_vol_discr}
+
+    # Define boundary connections
+    dist_boundary_connections = set_up_distributed_communication(
+        array_context, mesh,
+        volume_discr,
+        discr_tag_to_group_factory, comm=mpi_communicator
+    )
+
+    return DiscretizationCollection(
+        array_context=array_context,
+        mesh=mesh,
+        discr_tag_to_group_factory=discr_tag_to_group_factory,
+        discr_from_dd=discr_from_dd,
+        dist_boundary_connections=dist_boundary_connections,
+        mpi_communicator=mpi_communicator
+    )
+
+
+def set_up_distributed_communication(
+        array_context: ArrayContext, mesh: Mesh,
+        volume_discr,
+        discr_tag_to_group_factory, comm=None) -> dict:
```

Comment:
`
```suggestion
        discr_tag_to_group_factory, comm=None) -> Dict[key_type, value_type]:
```

Reply:
Snuck this in https://github.com/inducer/grudge/pull/122/commits/6bdc25ca42df6731fc2f632403f1a64d64c75f28